### PR TITLE
Restructuring the code

### PR DESCRIPTION
**Pull request by rrtoledo**
## Content
Most of the code is inside a single file centralized_telescope.rs. This PR intends to split it in different files within a new folder centralized_telescope.

## Pre-submit checklist

- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
- PR
    - [ ] No clippy warnings in the CI
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- Documentation
    - [ ] Update README file (if relevant)
    - [ ] Update documentation website (if relevant)

## Comments

I have also created a new utils folder to put the shared types for all Alba algorithms, i.e. `Element` so far. 

## Issue(s)

Relates to #46

### CODE REVIEW

**Review by tolikzinovyev — COMMENTED** [review]
Notes from a call with @rrtoledo:
`params.rs`, `setup.rs`, `proof.rs` - structs without functions, `pub` member variables
`make_setup.rs` - function converting `Params` to `Setup` + auxiliary functions
`algorithms.rs` - `proof()`, `verify()` with some or all auxiliary functions
`mod.rs` - just submodule declarations
move `centralized_telescope/utils.rs` to `utils/sample.rs` or similar since simple lottery will need it
`round.rs` can stay as is or moved into `algorithms.rs`

**Review by tolikzinovyev — COMMENTED** [review]
Thanks for the PR! Is there a reason to keep `impl` for `Setup` and `Proof`? I thought we were going to have free functions like
```
func MakeSetup(params: &Params) -> Setup {
}
```

**Comment by rrtoledo** [comment]
> Thanks for the PR! Is there a reason to keep `impl` for `Setup` and `Proof`? I thought we were going to have free functions like
> 
> ```
> func MakeSetup(params: &Params) -> Setup {
> }
> ```

No reasons. I forgot we discussed about removing the impl. Will update it now.

File: src/centralized_telescope/mod.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//! ALBA's bounded DFS scheme using Blake2b as hash function.
+//! (c.f. Section 3.2.2 of Alba paper)
+
+// Prove and Verify functions
+pub mod algorithm;
+
+// Setup function
+pub mod init;
+
+// Structures
```
Comment (tolikzinovyev) [inline src/centralized_telescope/mod.rs:10]: nit: it wasn't super clear that this applies to the four items below, maybe remove empty lines in between?

File: src/centralized_telescope/params.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+//! ALBA's Setup structure
+
+/// Setup input parameters
+#[derive(Debug, Clone, Copy)]
+pub struct Params {
+    /// Soundness security parameter
+    pub(super) lambda_sec: f64,
```
Comment (tolikzinovyev) [inline src/centralized_telescope/params.rs:7]: as discussed, these variable should be `pub`

File: src/centralized_telescope/setup.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+//! ALBA's Setup structure comprising among others numbers of elements and internal parameters
```
Comment (tolikzinovyev) [inline src/centralized_telescope/setup.rs:1]:
```suggestion
//! ALBA's Setup structure comprising among others number of elements (proof size) and internal parameters
```

**Review by tolikzinovyev — COMMENTED** [review]
This looks very good, thanks! Left comments for the remaining smaller issues.

**Review by djetchev — APPROVED** [review]
lgtm

File: src/centralized_telescope/round.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        hasher.update(b"Telescope-H1");
+        hasher.update(first_input);
+        hasher.update(second_input);
+        let digest: Hash = hasher.finalize().into();
+        (digest, utils::sample_uniform(&digest, n_p))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub(super) fn new(v: u64, t: u64, n_p: u64) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/centralized_telescope/round.rs:28]: It is at the top! I guess this is an old comment

File: src/centralized_telescope/setup.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            u: u_f64 as u64,
+            r: (params.lambda_rel / lambda_rel1).ceil() as u64,
+            d: d as u64,
+            q: 2.0 * lbar_over_d,
+            b: ((w * lbar_over_d + 1.0) * exponential * d * u_f64 + d).floor() as u64,
+        }
+    }
+
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
```
[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/centralized_telescope/setup.rs]: Likewise, I think this is an old comment. It is at the top of the  file

**Review by tolikzinovyev — APPROVED** [review]
LGTM

**Pull request merged**